### DISCUSSION TITLE
How to use 7 segment common anode display to display all sensor values rotating on a timer

### DISCUSSION THREAD

**Discussion by HarHarLinks** [arendst/Tasmota#14451] [Q&A] [answered]
I have a BME280 (fully working) and a 7 segment common anode display over I2C hooked up to Tasmota 10.1.

The display was also autodetected. According to its label it's a 4 digit 0.56" 7-Segment HT16K33 backpack by Adafruit that does the I2C decoding and drives the display. Additionally to the 7 segments, it supports a `:` in the middle (e.g. can display time) as well as `.` to display floats.

`DisplayMode 1` to display time works.
`DisplayMode 0` and using `DisplayText` works only on (signed) integers. Floats will be cut off to integers, Text won't display at all. I have not found a command to manually set the LED bitmasks yet to write text myself.

I would like to display time (24h format) and sensor values on a loop, e.g.
1. display time for 5 seconds
2. display temperature value with unit for 5 seconds
3. display humidity value with unit for 5 secs
4. display pressure value without unit (since it can be >999) for 5 secs

My build supports Rules and Timers, but I also have some Free Program Space 1248 kB and Free RAM 22.0 kB.

- I tried creating a rule with triggers on 5 seconds, but only whole minutes would work. That is too slow. How do I use seconds?
- How do I display different sensor values? And their Units?
- Which of these display types does mine belong to? https://tasmota.github.io/docs/Displays/#displaymode Please update the docs!
- How do I write Text or custom bits to my display?
- If I can use Rules for the time triggers, how do I access the sensor values in the command?

**Comment by lalo-uy** [comment]
In a similar user case I display 2 DS18B20 sensors every 5 sec.
Sensor 2 is rounded to int, but sensor 1 shows 1 decimal digit.
^ show as C on the display

Rule1
ON Tele-DS18B20-2#Temperature DO  Backlog
      Var1 %value% ;
      ADD1 0.5 ;
ENDON
ON Tele-DS18B20-1#Temperature DO  Backlog
      DisplayText %value%^ ;
      RuleTimer1 5
ENDON
ON rules#timer=1 DO Backlog
      DisplayNumber %Var1%,1,0,3 ;
      DisplayTextNC  ^
ENDON

**Reply by HarHarLinks** [reply]
There is an example that used `Tele-` as triggers in the docs. However all values on my BME280 are updated simultaneously, so only one would ever get displayed. `RuleTimer` might be the solution I was looking for, I will try it out.
What do your `^` characters mean?

**Reply by sfromis** [reply]
If you want to to get sensor data at shorter intervals than TelePeriod, do not use the `tele-` prefix. Then the triggers will will fire every few seconds, time depends on the sensor driver. You can then use rules to pick up each value to separate variables, and then rotate what you use in your display rules timed by RuleTimer.

**Reply by barbudor** [reply]
> all values on my BME280 are updated simultaneously, so only one would ever get displayed.

They are all separted in the JSON. You can have a trigger for each and display each on with proper cusor positionning
```
  ON tele-BME280#Temperature DO DisplayText [....]Temp: %value% C ENDON
  ON tele-BME280#Humidity DO DisplayText [....]Hum: %value% %% ENDON
  ON tele-BME280#Pressure DO DisplayText [...]Pres: %value% hPa ENDON
```
Replace the `....`  by the correct DisplayText commands to position the cursor as needed

**Reply by HarHarLinks** [reply]
> If you want to to get sensor data at shorter intervals than TelePeriod

(How) can I customize the interval at which tasmota reads the sensor values? It seems to do so several times per second (if I set `DisplayMode 2`), which is totally unnecessary and might even heat up the sensor unnecessarily. Every 5 to 60 seconds would be precise enough.

**Reply by sfromis** [reply]
You *cannot* customize reading interval, what you can do is to get all the sensor values at the same time, and then use RuleTimer to control  your display schedule. What barbudor posted could only work if you had a display with room for all 3 values at the same time.

**Comment by lalo-uy** [comment]
^  as the last character in
  DisplayText %value%^ ;
Shows up as ° for degrees
If BEM loads all values at the same time you need to save them on Vars to
later display.

**Comment by lalo-uy** [comment, marked as answer]
Since there 4 values showing 5 sec, set up Teleperios 20
Then
ON tele-BME280#Temperature DO  Var1 %value%  ENDON
ON tele-BME280#Humidity DO Var2 %value%  ENDON
ON tele-BME280#Pressure DO Backlog
      DisplayText %value% ;
      RuleTimer1 5
      RuleTimer2 10
      RuleTimer3 15
ENDON
ON rules#timer=1 DO DisplayText Var1 ENDON
ON rules#timer=2 DO DisplayText Var2 ENDON
ON rules#timer=3 DO DisplayText ...(timeas you like).. ENDON

**Reply by HarHarLinks** [reply]
Yes! Thank you, this should in principle work:
```
Rule1
ON BME280#Temperature DO backlog
var1 %value%;
ENDON
ON BME280#Humidity DO backlog
var2 %value%;
ENDON
ON BME280#Pressure DO backlog
var3 %value%;
ENDON
ON Time#Minute|1 DO backlog
DisplayMode 2;
RuleTimer1 5;
ENDON
ON rules#timer=1 DO backlog
DisplayMode 0;
DisplayFloat %var1%,1,1,3;
DisplayTextNC ^;
RuleTimer2 5;
ENDON
ON rules#timer=2 DO backlog
DisplayFloat %var2%,1,1,3;
DisplayTextNC H;
RuleTimer3 5;
ENDON
ON rules#timer=3 DO backlog
DisplayNumber %var3%,1,0,3;
DisplayTextNC P;
RuleTimer4 5;
ENDON
ON rules#timer=4 DO backlog
DisplayMode 2;
RuleTimer1 5;
ENDON
```

`DisplayMode 2` for my display is a 24h-clock with the `:` blinking every second, which is nice.

Now my only problem is that some of these DisplayText, DisplayFloat, DisplayNumber commands do not work properly. I can not write any characters to the display at all, see my other comment.

**Reply by HarHarLinks** [reply]
This kind of display uses a completely different way to DisplayText that is practically undocumented. Find more here: https://github.com/arendst/Tasmota/pull/8900

Find my final Rule "program" using above special formatting commands here: https://gist.github.com/HarHarLinks/e06363c4b61d6be2f0941193bacf1879

**Comment by HarHarLinks** [comment]
Docs say: https://tasmota.github.io/docs/Commands/#displays

DisplayTextNC: Clears first, then displays text
DisplayNumberNC: Display integer number as above, but without clearing first

which is wrong?

**Reply by pkkrusty** [reply]
NC doesn't clear the display. I fixed one description but didn't notice the other until yesterday.

**Comment by lalo-uy** [comment]
Sweet the display style  you like the most.
You can specify position and number for digits as DisplayNumber
%Var1%,1,0,3 ;

**Reply by HarHarLinks** [reply]
DisplayNumber command in the console does not work at all. For example I tried:
```
16:06:15.245 CMD: DisplayNumber 123
16:06:15.251 RSL: RESULT = {"DisplayNumber":"123"}
16:06:19.497 CMD: DisplayNumber 1234
16:06:19.503 RSL: RESULT = {"DisplayNumber":"1234"}
```
but display does not update. I also tried
```
16:05:36.980 CMD: displaytext 1234,0,4
16:05:36.989 RSL: RESULT = {"DisplayText":"1234,0,4"}
16:05:41.732 CMD: displaytext 123,0,4
16:05:41.741 RSL: RESULT = {"DisplayText":"123,0,4"}
```
Here 123 should be aligned to the left: "position can be 0 (left-most) to NUM_DIGITS-1 (right-most)".
But it does not respect the position argument.

```
16:08:49.550 CMD: displaytext ^
16:08:49.558 RSL: RESULT = {"DisplayText":"^"}
```
I found this explained in the docs, but it does not work for me, the display does not update.

`DisplayRaw` does not work either.

**Reply by HarHarLinks** [reply]
How to use it is not documented, but I found it here: https://github.com/arendst/Tasmota/pull/8900

**Comment by lalo-uy** [comment]
I remember some tweaks in the initialization sequence of DisplayMode to
make it work.
Don't forget  DisplayModel 15 also

**Reply by HarHarLinks** [reply]
```
16:39:22.076 CMD: DisplayModel 15
16:39:22.082 RSL: RESULT = {"DisplayModel":11}
```

**Comment by lalo-uy** [comment]
My experience is with TM1637, TM1638 and MAX7219 Seven-Segment Displays
You driver could be difernet.
